File: example/src/main/scala/com/github/j5ik2o/pekko/persistence/effector/example/scalaimpl/defaultStyle/nonPersistence/oop/BankAccountAggregate.scala
package com.github.j5ik2o.pekko.persistence.effector.example.scalaimpl.defaultStyle.nonPersistence.oop

import com.github.j5ik2o.pekko.persistence.effector.example.scalaimpl.*
import com.github.j5ik2o.pekko.persistence.effector.example.scalaimpl.defaultStyle.nonPersistence.{
  BankAccount,
  BankAccountAggregateState,
}
import org.apache.pekko.actor.typed.Behavior
import org.apache.pekko.actor.typed.scaladsl.{AbstractBehavior, ActorContext, Behaviors}

object BankAccountAggregate {
  def actorName(aggregateId: BankAccountId): String =
    s"${aggregateId.aggregateTypeName}-${aggregateId.asString}"

  def apply(
    id: BankAccountId,
  ): Behavior[BankAccountCommand] =
    Behaviors.setup { implicit ctx =>
      new BankAccountAggregate(id, BankAccountAggregateState.NotCreated(id))(ctx)
    }
}

final class BankAccountAggregate(id: BankAccountId, initialState: BankAccountAggregateState)(
  ctx: ActorContext[BankAccountCommand],
) extends AbstractBehavior[BankAccountCommand](ctx) {

  private var bf = initialState match {
    case state: BankAccountAggregateState.NotCreated =>
      notCreated(state)
    case state: BankAccountAggregateState.Created =>
      created(state)
  }

  override def onMessage(msg: BankAccountCommand): Behavior[BankAccountCommand] =
    bf(msg)

  private def notCreated(state: BankAccountAggregateState.NotCreated)
    : PartialFunction[BankAccountCommand, Behavior[BankAccountCommand]] = {
    case BankAccountCommand.Create(aggregateId, limit, balance, replyTo) =>
      val Result(bankAccount, event) = BankAccount.create(aggregateId, limit, balance)
      replyTo ! CreateReply.Succeeded(aggregateId)
      val newState: BankAccountAggregateState.Created =
        BankAccountAggregateState.Created(state.aggregateId, bankAccount)
      bf = created(newState)
      this
  }

  private def created(state: BankAccountAggregateState.Created)
    : PartialFunction[BankAccountCommand, Behavior[BankAccountCommand]] = {
    case BankAccountCommand.Stop(aggregateId, replyTo) =>
      replyTo ! StopReply.Succeeded(aggregateId)
      Behaviors.stopped
    case BankAccountCommand.GetBalance(aggregateId, replyTo) =>
      replyTo ! GetBalanceReply.Succeeded(aggregateId, state.bankAccount.balance)
      Behaviors.same
    case BankAccountCommand.DepositCash(aggregateId, amount, replyTo) =>
      state.bankAccount
        .add(amount)
        .fold(
          error => {
            replyTo ! DepositCashReply.Failed(aggregateId, error)
            Behaviors.same
          },
          { case Result(newBankAccount, event) =>
            replyTo ! DepositCashReply.Succeeded(aggregateId, amount)
            bf = created(state.copy(bankAccount = newBankAccount))
            this
          },
        )
    case BankAccountCommand.WithdrawCash(aggregateId, amount, replyTo) =>
      state.bankAccount
        .subtract(amount)
        .fold(
          error => {
            replyTo ! WithdrawCashReply.Failed(aggregateId, error)
            Behaviors.same
          },
          { case Result(newBankAccount, event) =>
            replyTo ! WithdrawCashReply.Succeeded(aggregateId, amount)
            bf = created(state.copy(bankAccount = newBankAccount))
            this
          },
        )
    case _ =>
      Behaviors.unhandled
  }

}

File: example/src/main/java/com/github/j5ik2o/pekko/persistence/effector/example/javaimpl/defaultStyle/nonPersistence/oop/BankAccountAggregate.java
package com.github.j5ik2o.pekko.persistence.effector.example.javaimpl.defaultStyle.nonPersistence.oop;

import com.github.j5ik2o.pekko.persistence.effector.example.javaimpl.BankAccountCommand;
import com.github.j5ik2o.pekko.persistence.effector.example.javaimpl.BankAccountId;
import com.github.j5ik2o.pekko.persistence.effector.example.javaimpl.CommandReply;
import org.apache.pekko.actor.typed.Behavior;
import org.apache.pekko.actor.typed.javadsl.AbstractBehavior;
import org.apache.pekko.actor.typed.javadsl.ActorContext;
import org.apache.pekko.actor.typed.javadsl.Behaviors;
import org.apache.pekko.actor.typed.javadsl.Receive;

public class BankAccountAggregate extends AbstractBehavior<BankAccountCommand> {
  private final BankAccountAggregateState state;

  public static String actorName(BankAccountId aggregateId) {
    return aggregateId.getAggregateTypeName() + "-" + aggregateId.asString();
  }

  public static Behavior<BankAccountCommand> create(BankAccountId aggregateId) {
    return Behaviors.setup(
        ctx ->
            new BankAccountAggregate(ctx, new BankAccountAggregateState.NotCreated(aggregateId)));
  }

  private BankAccountAggregate(
      ActorContext<BankAccountCommand> context, BankAccountAggregateState state) {
    super(context);
    this.state = state;
  }

  @Override
  public Receive<BankAccountCommand> createReceive() {
    return newReceiveBuilder()
        .onMessage(BankAccountCommand.Create.class, this::onCreate)
        .onMessage(BankAccountCommand.DepositCash.class, this::onDepositCash)
        .onMessage(BankAccountCommand.WithdrawCash.class, this::onWithdrawCash)
        .build();
  }

  private Behavior<BankAccountCommand> onWithdrawCash(
      BankAccountCommand.WithdrawCash withdrawCash) {
    if (state
        instanceof
        BankAccountAggregateState.Created(BankAccountId aggregateId, BankAccount bankAccount)) {
      var result = bankAccount.subtract(withdrawCash.amount());
      if (result.isRight()) {
        withdrawCash
            .replyTo()
            .tell(CommandReply.WithdrawCashReply.succeeded(aggregateId, withdrawCash.amount()));
        BankAccountAggregateState newState =
            new BankAccountAggregateState.Created(aggregateId, bankAccount);
        return new BankAccountAggregate(getContext(), newState);
      } else {
        withdrawCash
            .replyTo()
            .tell(CommandReply.WithdrawCashReply.failed(aggregateId, result.getLeft()));
        return this;
      }
    }
    return Behaviors.unhandled();
  }

  private Behavior<BankAccountCommand> onDepositCash(BankAccountCommand.DepositCash depositCash) {
    if (state
        instanceof
        BankAccountAggregateState.Created(BankAccountId aggregateId, BankAccount bankAccount)) {
      var result = bankAccount.add(depositCash.amount());
      if (result.isRight()) {
        depositCash
            .replyTo()
            .tell(CommandReply.DepositCashReply.succeeded(aggregateId, depositCash.amount()));
        BankAccountAggregateState newState =
            new BankAccountAggregateState.Created(aggregateId, bankAccount);
        return new BankAccountAggregate(getContext(), newState);
      } else {
        depositCash
            .replyTo()
            .tell(CommandReply.DepositCashReply.failed(aggregateId, result.getLeft()));
        return this;
      }
    }
    return Behaviors.unhandled();
  }

  private Behavior<BankAccountCommand> onCreate(BankAccountCommand.Create createCommand) {
    if (state instanceof BankAccountAggregateState.NotCreated) {
      BankAccount bankAccount = BankAccount.create(createCommand.aggregateId());
      createCommand.replyTo().tell(CommandReply.CreateReply.succeeded(createCommand.aggregateId()));
      BankAccountAggregateState newState =
          new BankAccountAggregateState.Created(createCommand.aggregateId(), bankAccount);
      return new BankAccountAggregate(getContext(), newState);
    }
    return Behaviors.unhandled();
  }
}
